src/Domain/FunctionsOfTime/ThreadsafeList.hpp

```diff
@@ -0,0 +1,317 @@
+  interval_list_ = std::move(other.interval_list_);
+  most_recent_interval_.store(interval_list_.get(), std::memory_order_release);
+  other.most_recent_interval_.store(nullptr, std::memory_order_release);
+  return *this;
+}
+
+template <typename T>
+auto ThreadsafeList<T>::operator=(const ThreadsafeList& other)
+    -> ThreadsafeList& {
+  if (this == &other) {
+    return *this;
+  }
+  initial_time_ = other.initial_time_;
+
+  std::unique_ptr<Interval>* previous_pointer = &interval_list_;
+  for (auto&& entry : other) {
+    previous_pointer->reset(
+        new Interval{entry.expiration, entry.data, nullptr});
+    previous_pointer = &(*previous_pointer)->previous;
+  }
+
+  most_recent_interval_.store(interval_list_.get(), std::memory_order_release);
+  return *this;
+}
+
+template <typename T>
+void ThreadsafeList<T>::insert(const double update_time, T data,
+                               const double expiration_time) {
+  const auto* old_interval =
+      most_recent_interval_.load(std::memory_order_acquire);
+  const double old_expiration =
+      old_interval != nullptr ? old_interval->expiration : initial_time_;
+  if (old_expiration != update_time) {
+    ERROR("Tried to insert at time "
+          << update_time << ", which is not the old expiration time "
+          << old_expiration);
+  }
+  if (expiration_time <= update_time) {
+    ERROR("Expiration time " << expiration_time << " is not after update time "
+                             << update_time);
+  }
+  std::unique_ptr<Interval> new_interval(new Interval{
+      expiration_time, std::move(data), std::move(interval_list_)});
+  const auto* const new_interval_p = new_interval.get();
+  interval_list_ = std::move(new_interval);
+  if (not most_recent_interval_.compare_exchange_strong(
+          old_interval, new_interval_p, std::memory_order_acq_rel)) {
+    ERROR("Attempt at concurrent modification detected.");
+  }
+}
+
+template <typename T>
+auto ThreadsafeList<T>::operator()(const double time) const -> IntervalInfo {
+  if (time < initial_time_ and not equal_within_roundoff(time, initial_time_)) {
+    ERROR("Requested time " << time << " precedes earliest time "
+                            << initial_time_);
+  }
+  const auto& interval = find_interval(time, false);
+  const double start = interval.previous != nullptr
+                           ? interval.previous->expiration
+                           : initial_time_;
+  return {start, interval.data, interval.expiration};
+}
+
+template <typename T>
+double ThreadsafeList<T>::expiration_time() const {
+  auto* interval = most_recent_interval_.load(std::memory_order_acquire);
+  return interval != nullptr ? interval->expiration : initial_time_;
+}
+
+template <typename T>
+double ThreadsafeList<T>::expiration_after(const double time) const {
+  return find_interval(time, true).expiration;
+}
+
+template <typename T>
+void ThreadsafeList<T>::pup(PUP::er& p) {
+  p | initial_time_;
+  p | interval_list_;
```

Comment:
This needs to be fixed to be threadsafe.

Reply:
I think it's safe to assume that if we are serializing or deserializing, we won't be updating/inserting. The only operation that has to be threadsafe is sizing since that can be done at any time. However, we should discuss this to confirm.

Reply:
We serialize during IO.
https://github.com/sxs-collaboration/spectre/blob/04376d11fdd83829d72508ca1db9f7674fb0c148/src/IO/Observer/VolumeActions.hpp#L378